Move documentation to doc/ folder and fill some gaps in the documentation (#1154)
mono/linker: docs/design/steps-explained.md

# Inside the linker

The linker is a quite small piece of code, and it pretty simple to address.
Its only dependency is `Mono.Cecil`, that is used to read, modify and write back
the assemblies.

Everything is located in the namespace Linker, or in sub namespaces.
Being a command line utility, its entry point function is in the class Driver.

> This class is in charge of analyzing the command line, and to instantiate two

Reviewer: ```suggestion
This class is in charge of analyzing the command line, and instantiating two
```

important objects, a LinkContext, and a Pipeline.

The LinkContext contains all the information that will be used during the
linking process, such as the assemblies involved, the output directory and

> probably other useful stuff.

Reviewer: ```suggestion
other data shared across steps.
```

The Pipeline is simply a queue of actions (steps), to be applied to the current
context. The whole process of linking is split into those different steps
that are all located in the Linker.Steps namespace.

Here are the current steps that are implemented, in the order they are used:

## ResolveFromAssembly or ResolveFromXml

Those steps are used to initialize the context and pre-mark the root code
that will be used as a source for the linker.

Resolving from an assembly or resolving from an xml descriptor is a decision
taken in the command line parsing.

## LoadReferences

This step will load all the references of all the assemblies involved in the
current context.

## Blacklist

This step is used if and only if you have specified that the core should be
linked. It will load a bunch of resources from the assemblies, that are
actually a few XML descriptors, that will ensure that some types and methods

> that are used from inside the runtime are properly linked and not removed.

Reviewer: Maybe mention the substitution XML here too.

It is doing so by inserting a ResolveFromXml step per blacklist in the
pipeline.

## Mark

This is the most complex step. The linker will get from the context the list
of types, fields and methods that have been pre-marked in the resolve steps,
and walk through all of them. For every method, it will analyse the CIL stream,
to find references to other fields, types, or methods.

When it encounters such a reference, it will resolve the original definition of
this reference, and add this to the queue of items to be processed. For

> instance, if have in a source assembly a call to Console.WriteLine, the linker

Reviewer: ```suggestion
instance, if a source assembly has a call to Console.WriteLine, the linker
```

will resolve the appropriate method WriteLine in the Console type from the

> mscorlib assembly, and add it to the queue. When this WriteLine method will be

Reviewer: ```suggestion
mscorlib assembly, and add it to the queue. When this WriteLine method is
```

dequeued, and processed, the linker will go through everything that is used in

> it, and add it to the queue, if they have not been processed already.

Reviewer 2: ```suggestion
it, and add it to the queue, if it hasn't been processed already.
```

> it, and add it to the queue, if they have not been processed already.

Reviewer: ```suggestion
it, and add it to the queue, if it has not been processed already.
```

To know if something has been marked to be linked, or processed, the linker
is using a functionality of Cecil called annotations. Almost everything in

> Cecil can be annotated. Concretely, it means that almost everything owns an

Reviewer 2: ```suggestion
Cecil can be annotated. Specicialy, it means that almost everything owns a
```

> Hashtable in which you can add what you want, using the keys and the values you

Reviewer 2: ```suggestion
hashtable in which you can add what you want, using the keys and the values you
```

want.

So the linker will annotate assemblies, types, methods and fields to know

> what should be linked or not, and what has been processed, and how it should

Reviewer 2: ```suggestion
what should be linked or not, what has been processed and how it should
```

process them.

This is really useful as we don't have to recreate a full hierarchy of classes
to encapsulate the different Cecil types to add the few pieces of information we want.

## Sweep

This simple step will walk through all the elements of an assembly, and based
on their annotations, remove them or keep them.

## Clean

This step will clean parts of the assemblies, like properties. If a property

> used to have a getter and a setter, and that after the mark & sweep steps,
> only the getter is linked, it will update the property to reflect that.

Reviewer 2: ```suggestion
used to have a getter and a setter and after the mark & sweep steps
only the getter is retained it will update the property to reflect that.
```
